src/bundles/ar/libraries/object_state_library/ARObject.tsx

```diff
@@ -0,0 +1,394 @@
+import {
+  BoxGeometry,
+  MeshStandardMaterial,
+  SphereGeometry,
+  Vector3,
+} from 'three';
+import {
+  type Behaviours,
+  LightModel,
+  RenderWithinDistance,
+  ShapeModel,
+  InterfaceModel,
+  type RotationClass,
+  parseRotation,
+  type RenderClass,
+  parseRender,
+  parseMovement,
+  type MovementClass,
+  GltfModel,
+} from './Behaviour';
+import ARObjectComponent from './ARObjectComponent';
+import { parseVector3 } from './Misc';
+import { parseJsonInterface } from './model_components/InterfaceComponent';
+
+/**
+ * Abstract class for an AR object.
+ * Extend this class to create your custom AR object.
+ */
+export class ARObject {
+  type: string = ''; // Unique identifier for class
+  id: string;
+  position: Vector3;
+  behaviours: Behaviours;
+  uuid: string | undefined = undefined;
+  isInFront = false;
+  isSelected = false;
+  onSelect?: (object: ARObject) => void;
+  constructor(
+    id: string,
+    position: Vector3,
+    behaviours: Behaviours,
+    onSelect?: (object: ARObject) => void,
+  ) {
+    this.id = id;
+    this.position = position;
+    this.behaviours = behaviours;
+    this.onSelect = onSelect;
+  }
+  toJSON = () => {
+    const object = { ...this };
+    const behavioursClone = { ...this.behaviours } as any;
+    delete behavioursClone.model;
+    object.behaviours = behavioursClone;
+    object.isInFront = false;
+    return object;
+  };
+  static fromObject = (object: any) => {
+    if (!object) return;
+    let isSelected = false;
+    const tempSelected = object.isSelected;
+    if (typeof tempSelected === 'boolean') {
+      isSelected = tempSelected;
+    }
+    let newObject = CubeObject.parseObject(object);
+    if (newObject) {
+      newObject.isSelected = isSelected;
+      return newObject;
+    }
+    newObject = SphereObject.parseObject(object);
+    if (newObject) {
+      newObject.isSelected = isSelected;
+      return newObject;
+    }
+    newObject = GltfObject.parseObject(object);
+    if (newObject) {
+      newObject.isSelected = isSelected;
+      return newObject;
+    }
+    newObject = UIObject.parseObject(object);
+    if (newObject) {
+      newObject.isSelected = isSelected;
+      return newObject;
+    }
+    newObject = LightObject.parseObject(object);
+    if (newObject) {
+      newObject.isSelected = isSelected;
+      return newObject;
+    }
```

Comment:
This could be simplified to 

```
    let newObject = CubeObject.parseObject(object) ||
                    SphereObject.parseObject(object) ||
                    GltfObject.parseObject(object) ||
                    UIObject.parseObject(object) ||
                    LightObject.parseObject(object);

    if (newObject) {
        newObject.isSelected = isSelected;
        return newObject;
    }
```